### DISCUSSION TITLE
Struggling with task 10

### DISCUSSION THREAD

**Discussion by jp868-cam** [Part-II-Computational-Physics/.github#47] [General]
I am having hard time understanding what to do in task 10. Is this the right method to find the expected value using sparse matrix method or should have I started with the sparse matrix from the beginning? Also, is there a way to check my code is doing what it's supposed to do?

    def ExpValueMx(psiIn):

       N =round( np.log(np.size(psiIn)) / np.log(2) )    
       Matrices1 = np.zeros((N, 2, 2))
       Matrices1[:] = sigmaX
 
       m_x = sparse.csr_matrix(ft.reduce(sparse.kron, Matrices1)/N)  # m_x operator
    
       expVal = np.inner(psiIn.T, m_x @ psiIn) # <psi| m_x |psi>
       return expVal

**Comment by JNHallen** [comment]
You are on the right track, but you are making some minor mistakes in the code. Remember that the magnetisation $m^x$ is defined as 
$$\frac{1}{N} \sum_i I_1...I_{i-1}\sigma_i^x I_{i+1}...I_N$$
If you want to construct the matrix version of the operator using the method you have attempted above, you will need to use 
```
Matrices1 = np.zeros((N, 2, 2))
Matrices1[:] = I
Matrices1[0] = sigmaX
```
and then do a sum like in CompleteHamiltonian, using the Kronecker product and the np.roll function. 

If things are working properly you should find $\langle m^x \rangle = 0$ for $h \ll 1$ and  $\langle m^x \rangle = 1$ for $h \gg 1$.

**Reply by jp868-cam** [reply]
Thank you!
